orc/orcarm.c

```diff
@@ -1016,3 +1016,211 @@ orc_arm64_emit_am (OrcCompiler *p, OrcArm64RegBits bits, OrcArm64DP opcode,
 
   orc_arm_emit (p, code);
 }
+
+/** data processing instructions: Logical
+ *
+ * Immediate
+ *    3                   2                   1
+ *  1 0 9 8 7 6 5 4 3 2 1 0 9 8 7 6 5 4 3 2 1 0 9 8 7 6 5 4 3 2 1 0
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |b|op |1 0 0 1 0 0|         imm13           |    Rn   |    Rd   |
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ *
+ * Shifted register
+ *    3                   2                   1
+ *  1 0 9 8 7 6 5 4 3 2 1 0 9 8 7 6 5 4 3 2 1 0 9 8 7 6 5 4 3 2 1 0
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |b|op |0 1 0 1 0|sft|0|    Rm   |   imm6    |    Rn   |    Rd   |
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ */
+
+#define arm64_code_logical_imm(b,op,imm,Rn,Rd) (0x12000000 | \
+    (((b==64)&0x1)<<31) | (((op)&0x3)<<29) | (((imm)&0x1fff)<<10) | \
+    (((Rn)&0x1f)<<5)  | ((Rd)&0x1f))
+
+#define arm64_code_logical_reg(b,op,sft,Rm,imm,Rn,Rd) (0x0a000000 | \
+    (((b==64)&0x1)<<31) | (((op)&0x3)<<29) | (((sft)&0x3)<<22) | \
+    (((Rm)&0x1f)<<16) | (((imm)&0x3f)<<10) | (((Rn)&0x1f)<<5)  | ((Rd)&0x1f))
+
+#define mask_ones(val) ((val) && (((val) + 1) & (val)) == 0)
+#define mask_shifted_ones(val) ((val) && mask_ones(((val) - 1) | (val)))
+
+#if defined(__GNUC__)
+#define count_trailing_zeros(val) __builtin_ctzll(val)
+#define count_leading_zeros(val) __builtin_ctzll(val)
```

Comment:
count_leadng_zeros == count_trailing_zeros ?

Reply:
Oh.... it should be `__builtin_clzll`